API - application program interface
REST - representational state transfer

But in everyday parlance, a REST API just means a website that produces data intended for another computer program to consume, rather than something intended to be displayed to people in a browser.

Caching is a way to reduce the total number of calls that you have to make to the APIs. 

### URL 
http://umich.edu/about
1. http is our protocol, says how to communicate 
2. umcih.edu is our server or domain, where to
3. about is our argument or path, what to

url <scheme.>://<host.>:<port.>/<path.>

http(80) and https, ftp and mailto (for email addresses).
1. the default port number is 80
2. the host can be an IP address directly. not so common because a URL containing a domain name will continue to work even if the remote server keeps its domain name but moves to a different IP address.

HTTP is technically (previously) documented as IETF RFC 2616.

### ip and dns
IP stands for Internet Protocol. No two computers that are connected to the internet have the same IP address at the same time. total four chunks, each chunk is represented by 8bits, hence the whole ip address is also called 32bit address

DNS, the Domain Name System

Routers are just computers whose sole job is to receive these data packets and pass them on.

while using HTTPS (to send/recieve from umich.edu), the first thing that my computer would do is send some information back and forth to the si.umich.edu server that would establish encryption keys so that all of the rest of the communication would be encrypted

Request headers and
Response headers (meta data) 

### Request Respone cycle  - 
1. Client makes a request to the server by saying GET <path.>, path is path part of urlif request involves sending some data, message starts with POST
(cookies involved)

2. the server responds to the client.
        1.The server first sends back some HTTP headers.
        2. A description of the type of content it is sending back
        3. Any cookies it would like the client to hold onto and send back the next time it communicates with the server.

### API 
Request to an API is made by visiting a URL. Some APIs use a different part of the HTTP protocol, called the post mechanism, in which case not all of the requests is encoded right into the URL. 

But we're going to focus only on the simplest mechanism, the HTTP get request, where the entire request is encoded in the URL.

### URL encoding

a URL path is not allowed to include the double -quote character. It also can’t include a : or / or a space. Whenever we want to include one of those characters in a URL, we have to encode them with other characters. A space is encoded as +. " is encoded as %22. : would be encoded as %3A. And so on.

In [2]:
import requests
import json

page = requests.get("https://api.datamuse.com/words?rel_rhy=funny")      
#Once we run requests.get, a python object is returned. 
#It’s an instance of a class called Response that is defined in the requests module.

print("page type",type(page))                             
print("page but only 150 char. ",page.text[:150])                  
#.text method, It contains the contents of the file or sometimes an error mssg
print("url ",page.url)      #print the url that was fetched
print("-"*30)
x = page.json()                                              
#turn page.text into a python object
#The .json() method. This converts the text into a python list or dictionary, 
#by passing the contents of the .text attribute to the jsons.loads function.

print(type(x))                                               #here, its a list of dictionaries
print("------first item in the list-------")
print(x[0])                                              #the first dictionary
print("------the whole list, pretty printed-------")
print(json.dumps(x, indent=2))                           # pretty print the results

#LITE
sums=0
for y in x:
    sums += y['numSyllables']
print(sums)

page  <class 'requests.models.Response'>
page but only 150 char.  [{"word":"money","score":4415,"numSyllables":2},{"word":"honey","score":1206,"numSyllables":2},{"word":"sunny","score":717,"numSyllables":2},{"word":"
url  https://api.datamuse.com/words?rel_rhy=funny
------------------------------
<class 'list'>
------first item in the list-------
{'word': 'money', 'score': 4415, 'numSyllables': 2}
------the whole list, pretty printed-------
[
  {
    "word": "money",
    "score": 4415,
    "numSyllables": 2
  },
  {
    "word": "honey",
    "score": 1206,
    "numSyllables": 2
  },
  {
    "word": "sunny",
    "score": 717,
    "numSyllables": 2
  },
  {
    "word": "bunny",
    "score": 702,
    "numSyllables": 2
  },
  {
    "word": "blini",
    "score": 613,
    "numSyllables": 2
  },
  {
    "word": "gunny",
    "score": 449,
    "numSyllables": 2
  },
  {
    "word": "tunny",
    "score": 301,
    "numSyllables": 2
  },
  {
    "word": "sonny",
    "score": 286,
    "numSyllables"

In the requests module, the get function is so smart that when it gets a 301, it looks at the new url and fetches it.

Example (not important though)
For example, github redirects all requests using http to the corresponding page using https (the secure http protocol). Thus, when we ask for http://github.com/presnick/runestone, github sends back a 301 code and the url https://github.com/presnick/runestone. 
The requests.get function then fetches the other url. It reports a status of 200 and the updated url. We have to do further inquire to find out that a redirection occurred (see below**).

The .headers attribute has as its value a dictionary consisting of keys and values. To find out all the headers, you can run the code and add a statement print(p.headers.keys()).

**
The .history attribute contains a list of previous responses, if there were redirects. 



The get function in the requests module takes an optional parameter called params. If a value is specified for that parameter, it should be a dictionary.

In [3]:
import requests
d = {'q': '"violins and guitars"', 'tbm': 'isch'}
results = requests.get("https://google.com/search", params=d)
print(results.url)

#we have another more illustrative example

https://www.google.com/search?q=%22violins+and+guitars%22&tbm=isch


In [5]:
import requests

# page = requests.get("https://api.datamuse.com/words?rel_rhy=funny")
kval_pairs = {'rel_rhy': 'funny'}
page = requests.get("https://api.datamuse.com/words", params=kval_pairs)
print(page.text[:10]) # print the first 150 characters
print(page.url) # print the url that was fetched

[{"word":"
https://api.datamuse.com/words?rel_rhy=funny


Q. If resp is a Response object returned by a call to requests.get(), which of the following is a way to extract the contents into a python dictionary or list?

A. resp.json()

B. json.loads(resp.text)

Ans. both

In [ ]:
#an example to get rhym-words for this

import requests

def get_rhymes(word):
    baseurl = "https://api.datamuse.com/words"
    params_diction = {} # Set up an empty dictionary for query parameters
    params_diction["rel_rhy"] = word
    params_diction["max"] = "3" # get at most 3 results
    resp = requests.get(baseurl, params=params_diction)
    # return the top three words
    word_ds = resp.json()
    return [d['word'] for d in word_ds]
    return resp.json() # Return a python object (a list of dictionaries in this case)

print(get_rhymes("funny"))


### Implementing requests_with_caching module
optimized this code for conceptual simplicity, so that it is useful as a teaching tool. It is not very efficient, because it always stores cached contents in a file, rather than saving it in memory. If you are ever implementing the caching pattern just for the duration of a program’s run, you might want to save cached content in a python dictionary in memory rather than writing it to a file.

The basic idea in the code is to maintain the cache as a dictionary with keys representing API requests that have been made, and values representing the text that was retrieved. In order to make our cache live beyond one program execution, we store it in a file. Hence, there are helper functions _write_to_file and read_to_file that write a cache dictionary to and read it from a file.

In [ ]:
import requests
import json

PERMANENT_CACHE_FNAME = "permanent_cache.txt"
TEMP_CACHE_FNAME = "this_page_cache.txt"

def _write_to_file(cache, fname):
    with open(fname, 'w') as outfile:
        outfile.write(json.dumps(cache, indent=2))      #neatly writes to outfile, whatever json is in cache 

def _read_from_file(fname):
    try:
        with open(fname, 'r') as infile:
            res = infile.read()              #txt from infile is written simply to res 
            return json.loads(res)           #same as res.json() which returns a python object created from the long text.
    except:
        return {}

def add_to_cache(cache_file, cache_key, cache_value):    
    temp_cache = _read_from_file(cache_file)                #temp_cache becomes a python object (dictionary)
    temp_cache[cache_key] = cache_value                    #writing the value under cache_key inside temp_cache dictionary
    _write_to_file(temp_cache, cache_file)                 #writes text in cachefile to tempcache

def clear_cache(cache_file=TEMP_CACHE_FNAME):             #clear
    _write_to_file({}, cache_file)

def make_cache_key(baseurl, params_d, private_keys=["api_key"]):     #default of private keys set to ["apikey"]
    """Makes a long string representing the query.
    Alphabetize the keys from the params dictionary so we get the same order each time.
    Omit keys with private info."""
    alphabetized_keys = sorted(params_d.keys())    #stores the keys in order
    res = []                                      #list
    for k in alphabetized_keys:
        if k not in private_keys:
            res.append("{}-{}".format(k, params_d[k]))    #stores the key which is not in private but in params_d.keys()
    return baseurl + "_".join(res)                 #returns url

def get(baseurl, params={}, private_keys_to_ignore=["api_key"], permanent_cache_file=PERMANENT_CACHE_FNAME, temp_cache_file=TEMP_CACHE_FNAME):
    full_url = requests.requestURL(baseurl, params)
    cache_key = make_cache_key(baseurl, params, private_keys_to_ignore)
    # Load the permanent and page-specific caches from files
    permanent_cache = _read_from_file(permanent_cache_file)
    temp_cache = _read_from_file(temp_cache_file)
    if cache_key in temp_cache:
        print("found in temp_cache")
        # make a Response object containing text from the change, and the full_url that would have been fetched
        return requests.Response(temp_cache[cache_key], full_url)
    elif cache_key in permanent_cache:
        print("found in permanent_cache")
        # make a Response object containing text from the change, and the full_url that would have been fetched
        return requests.Response(permanent_cache[cache_key], full_url)
    else:
        print("new; adding to cache")
        # actually request it
        resp = requests.get(baseurl, params)
        # save it
        add_to_cache(temp_cache_file, cache_key, resp.text)
        return resp

# Debugging calls to requests.get()

, you will not always get a Response object back from a call to requests.get. 

What you get back will generally be even more informative than what you get in the Runestone environment, but you have to know where to look.

1. The first thing that might go wrong is that you get a runtime error when you call requests.get(dest_url). There are two possibilities for what’s gone wrong in that case.
    One possibility is that the value provided for the params parameter is not a valid dictionary or doesn’t have key-value pairs that can be converted into text strings suitable for putting into a URL. 
        For example, if you execute requests.get("http://github.com", params = [0,1]), [0,1] is a list rather than a dictionary and the python interpreter generates the error, TypeError: 'int' object is not iterable.
        
    The second possibility is that the variable dest_url is either not bound to a string, or is bound to a string that isn’t a valid URL. 
        For example, it might be bound to the string "http://foo.bar/bat". foo.bar is not a valid domain name that can be resolved to an ip address, so there’s no server to contact. That will yield an error of type requests.exceptions.ConnectionError. Here’s a complete error message:
            `requests.exceptions.ConnectionError: HTTPConnectionPool(host='foo.bar', port=80): Max retries exceeded with url: /bat?key=val (Caused by <class 'socket.gaierror'>: [Errno 11004] getaddrinfo failed)`
            
The best approach is to look at the URL that is produced, eyeball it, and plug it into a browser to see what happens. Unfortunately, if the call to requests.get produces an error, you won’t get a Response object, so you’ll need some other way to see what URL was produced. 

The function defined below takes the same parameters as requests.get and returns the URL as a string, without trying to fetch it.

In [ ]:
import requests
def requestURL(baseurl, params = {}):
    # This function accepts a URL path and a params diction as inputs.
    # It calls requests.get() with those inputs,
    # and returns the full URL of the data you want to get.
    req = requests.Request(method = 'GET', url = baseurl, params = params)
    prepped = req.prepare()
    return prepped.url

print(requestURL(some_base_url, some_params_dictionary))


Assuming requestURL() returns a URL, match up what you see from the printout of the params dictionary to what you see in the URL that was printed out. 
If you have a sample of a URL from the API documentation, see if the structure of your URL matches what’s there. Perhaps you have misspelled one of the API parameter names or you misspelled the base url.

You can also try cutting and pasting the printed URL into a browser window, to see what error message you get from the website. 

You can then try changing the URL in the browser and reloading. When you finally get a url that works, you will need to translate the changes you made in the url back into changes to make to your baseurl or params dictionary.

If requests.get() executes without generating a runtime error, you are still not done with your error checking. No error means that your computer managed to connect to some web server and get some kind of response, but it doesn’t mean that it got the data you were hoping to get.

Fortunately, the response object returned by requests.get() has the .url attribute, which will help you with debugging.

It’s a good practice during program development to have your program print it out. This is easier than calling requestURL() but is only available to you if requests.get() succeeds in returning a Response object.

More importantly, you’ll want to print out the contents. Sometimes the text that’s retrieved is an error message that you can read, such as {"request empty": "There is no data that corresponds to your search."}. In other cases, it’s just obviously the wrong data. 
Print out the first couple hundred characters of the response text to see if it makes sense.


In [ ]:
import requests
dest_url = <some expr>
d = <some dictionary>
resp = requests.get(dest_url, params = d)
print(resp.url)
print(resp.text[:200])

Now you try it. Use requests.get() and/or requestURL() to generate the following url, https://www.google.com/search?tbm=isch&q=%22violins+and+guitars%22.

## Requests Cookbook
The basic process involves three steps:

1. Make the appropriate call to requests.get(). If you have trouble, print out the URL that’s generated and work with it in the browser.
2. Extract content from response object, by accessing the .text attribute and calling json.loads if the string is in json format.
3. Process the data you’ve extracted. Often, when you get back data in json format, it will be a highly nested data structure. You may only need a little of that data. You may want to review the chapter on nested data and nested iteration, especially the section on the cycle of Understand. Extract. Repeat.

The key to success is to make sure that you debug each of those steps before going on to the next one. This is just a particular case of the general advice we gave early in the course: start small and keep it working at every stage, growing the amount that your program does over time.

## Searching for media itunes

In [ ]:
#wont run bc with_caching thing
import requests_with_caching 
import json          #because after getting the data back from the api, its in json format

parameters = {"term": "Ann Arbor", "entity": "podcast"}       #figured this out with documentation at https://affiliate.itunes.apple.com/resources/documentation/itunes-store-web-service-search-api/
iTunes_response = requests_with_caching.get("https://itunes.apple.com/search", params = parameters, permanent_cache_file="itunes_cache.txt")
#the above line, gives back a response object, probably in json 

py_data = json.loads(iTunes_response.text) #the data, decoded into a python object,
for r in py_data['results']:            #figured this out with the structure of that python object
    print(r['trackName'])               #thats what we want, all the tracknames




## Searching for tags on flickr

https://www.flickr.com/services/api/
https://www.flickr.com/services/api/flickr.photos.search.html
at the documentation of the flickr api ->
at the bottom of this page, notice that they give what they call an endpoint. That's what we will call the base URL.
The next notice that they have a required parameter called the api_key.
A third thing to notice on this page is that they have another required query parameter called method.
We're going to have to say method equals flickr.photos.search. There are also a bunch of other keys that we can provide, what they call arguments.
we're going to search for tags.
Then, there's this tag_mode. If you provide more than one tag that can either be treated as you want to search for photos that have one of those tags as the or combination or that you want to search for photos that have all of those tags, that's the and combination. 
Most sites that provide XML also have some way to ask for JSON instead. Flickr does have weighed asked for JSON instead. They don't make it that easy to find out how.
Send a parameter called format with a value of JSON. So we're going to have in our URL format equals JSON.
There's one other little tricky thing. I'm just going to get rid of those markings, they have a little thing that says something about callback functions. It turns out we have to send nojsoncallback equals one.
Without that nojsoncallback equals one, you would get JSON results that are wrapped in basically a JavaScript function invocation called the callback. So that wrapping is part of a standard called JSONP. In any case, what we need to do is we need to include nojsoncallback equals one in order to just get pure JSON without any extra characters around it. 



base URL is https://api.flickr.com/services/rest/

key=value pairs, separated by &s:
1. One pair is method=flickr.photos.search. This says to do a photo search, rather than one of the many other operations that the API allows. Don’t be confused by the word “method” here– it is not a python method. That’s just the name flickr uses to distinguish among the different operations a client application can request.

2. format=json. This says to return results in JSON format.

3. per_page=5. This says to return 5 results at a time.

4. tags=mountains,river. This says to return things that are tagged with “mountains” and “river”.

5. tag_mode=all. This says to return things that are tagged with both mountains and river.

6. media=photos. This says to return photos

7. api_key=.... Flickr only lets authorized applications access the API. Each request must include a secret code as a value associated with api_key. Anyone can get a key. See the documentation for how to get one. We recommend that you get one so that you can test out the sample code in this chapter creatively. We have included some cached responses, and they are accessible even without an API key.

8. nojsoncallback=1. This says to return the raw JSON result without a function wrapper around the JSON response.




In [ ]:
# import statements
import requests_with_caching
import json
# import webbrowser

# apply for a flickr authentication key at http://www.flickr.com/services/apps/create/apply/?
# paste the key (not the secret) as the value of the variable flickr_key
flickr_key = 'yourkeyhere'

def get_flickr_data(tags_string):
    baseurl = "https://api.flickr.com/services/rest/"
    params_diction = {}
    params_diction["api_key"] = flickr_key # from the above global variable
    params_diction["tags"] = tags_string # must be a comma separated string to work correctly
    params_diction["tag_mode"] = "all"
    params_diction["method"] = "flickr.photos.search"
    params_diction["per_page"] = 5
    params_diction["media"] = "photos"
    params_diction["format"] = "json"
    params_diction["nojsoncallback"] = 1
    flickr_resp = requests_with_caching.get(baseurl, params = params_diction, permanent_cache_file="flickr_cache.txt")
    print(flickr_resp.url) # Paste the result into the browser to check what would it search if it didnt found it in the cache
    return flickr_resp.json()

result_river_mts = get_flickr_data("river,mountains")

# Some code to open up a few photos that are tagged with the mountains and river tags...

photos = result_river_mts['photos']['photo']
for photo in photos:
    owner = photo['owner']
    photo_id = photo['id']
    url = 'https://www.flickr.com/photos/{}/{}'.format(owner, photo_id)
    print(url)
    # webbrowser.open(url)


If this were a full Python environment, we would be able to import the web browser module, and we could call webbrowser.open, which would make this URL open up automatically in another tab.

# the final project
the base url ->
https://tastedive.com/api/similar
actuall usage ->
https://tastedive.com/api/similar?q=red+hot+chili+peppers%2C+pulp+fiction

parameters
1. q: the search query; consists of a atleat one name of movie(s separated by commas). Sometimes it is useful to specify the type of a certain resource in the query (e.g. if a movie and a book share the same title). You can do this by using the "band:", "movie:", for example "band:underworld, movie:harry potter. Don't forget to encode this parameter.
2. type: query type, specifies the desired type of results. just use type=movies in our case
3. info: when set to 1, additional information is provided for the recommended items, like a description and a related Youtube clip (when available). Default is 0.
4. limit: maximum number of recommendations to retrieve. Default is 20. 
5. k: Your API access key., we can keep it anything, we already have the data in the cache
6. callback: add when using JSONP, to specify the callback function.


The returned object contains, under the Similar key, the items that were searched for (a list in the Info key (key info != parameter info) and the recommended items (a list in the Results key). 
Each item in a list has the Name and Type keys. The type can be music, movie, show, book, author or game.

so the structure of the object is, a dictionary with one key (named 'similar'), in the value of that key, there is another dictionary with two keys (videlicet( formal for viz( namely/in other words)) Info and Results) , under both keys, theres a list, Each item in a list has the Name and Type keys 

In [ ]:
import requests_with_caching
import json

def get_movies_from_tastedive(name):
    parameters = {"q": name, "type": "movies","limit": 5}
    baseurl='https://tastedive.com/api/similar'
    movies_recom= requests_with_caching.get(baseurl, params = parameters)
    #print(movies_recom.url)
    return movies_recom.json()

def extract_movie_titles(names):
    alist=[]
    biglist=names['Similar']['Results']
    for dic in biglist:
        alist.append(dic['Name'])
    return alist

def get_related_titles(list1):
    blist=[]
    for name in list1:
        #print(name)
        blist += extract_movie_titles(get_movies_from_tastedive(name))
        #print(blist)
    
    #clearing duplicates
    blist = list(dict.fromkeys(blist))
    return blist

def get_movie_data(str1):
    parameters={'t':str1, 'r':'json'}
    baseurl='http://www.omdbapi.com/'
    resp_2=requests_with_caching.get(baseurl, params=parameters)
    #print(resp_2.url, resp_2.json())
    return resp_2.json()

def get_movie_rating(dict1):
    rate=0
    for rating in dict1['Ratings']:
        if rating['Source']=='Rotten Tomatoes':
            rate=int(rating['Value'].replace('%',''))
    return rate
        
def get_sorted_recommendations(alist):
    blist=get_related_titles(alist)
    #print(blist)
    dict1={}
    for movie in blist:
        rating=get_movie_rating(get_movie_data(movie))
        dict1[movie]=rating
    #print(dict1)
    slist = sorted(dict1, key=lambda w: (dict1[w], w), reverse=True)
    return slist
# some invocations that we use in the automated tests; uncomment these if you are getting errors and want better error messages
#get_sorted_recommendations(["Bridesmaids", "Sherlock Holmes"])



# Regular expression

In [3]:
import re 

TypeError: search() missing 1 required positional argument: 'string'

In [6]:
?re.search

In [8]:
print(re.search('[aeiou]', 'sky'))

None


In [11]:
re.findall('[aer]', 'are aer aera sfhush asjer')

['a', 'r', 'e']

In [ ]:
list(dict.fromkeys(re.findall('[aer]', 'are aer aera sfhush asjer')))


In [ ]:
line.find("from: ") ==0
line.startswith('from: ')
re.search('from: ', line)

## Socket module

In [13]:
import socket
mysock = socket.socket(socket.AF_INET, socket.SOCK_STREAM) 
mysock.connect('data.pr4e.org')     #443 https  23telnetlogin   22ssh(secure login)    21ftp
#by http, client always sends first request
cmd = 'GET http://data.pr4e.org/reo.txt HTTP/1.0\r\n\r\n'.encode()  #\r for nertworking, two times to disting. from headers
mysock.send(cmd)
while true: 
    data = mysock.recv(512)  #first 512 bits
    if (len(data)<1):
        break
    print(data.decode())   #decode converts incoming utf8 to strings(unicode)
mysock.close()


TypeError: getsockaddrarg: AF_INET address must be tuple, not str

In [ ]:
import urllib.request, urllib.parse, urllib.error 
fhand = urllib.request.urlopen('http....')
for line in fhand:
    print(line.decode().strip())
#does not prints headers

## beautiful soup (for a web scrapper)

In [ ]:
import urllib.request, urllib.parse, urllib.error 
import BeautifulSoup as bs
url = input('')
html = urllib.request.urlopen(url).read()
soup = bs(html, 'html.parser')
tags = soup('a')  #retrives all the anchor tags in the page 
for tag in tags:
    print(tag.get('href', None))
    

## JSON

In [ ]:
import json
data = '''{
    "name":"mahavir"
    "phone:"2914189284"
    "email":"f201902/"
}'''
info = json.loads(data)    #info is a python object, dict or list, data is a string 
print('Name:', info["Name"])
print('mail:', info["email"])


### google API with google maps

In [ ]:
import urllib.request, urllib.parse, urllib.error
import json
import ssl

api_key = False
# If you have a Google Places API key, enter it here
# api_key = 'AIzaSy___IDByT70'
# https://developers.google.com/maps/documentation/geocoding/intro

if api_key is False:
    api_key = 42
    serviceurl = 'http://py4e-data.dr-chuck.net/json?'                                #?
else :
    serviceurl = 'https://maps.googleapis.com/maps/api/geocode/json?'

# Ignore SSL certificate errors
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

while True:
    address = input('Enter location: ')
    if len(address) < 1:    continue                         #if we press enter instead of entering the location, it asks again

    parms = dict()
    parms['address'] = address
    if api_key is not False:
        parms['key'] = api_key                           #adding a element key to parms, read the above if else block

    url = serviceurl + urllib.parse.urlencode(parms)       #making the url, urlencode does what i have a screenshot of in c:, +means space, 2%C means comma

    print('Retrieving', url)
    uh = urllib.request.urlopen(url, context=ctx)             #just the handle of url
    data = uh.read().decode()                                 #actual string after decoding it from utf8 to unicode
    print('Retrieved', len(data), 'characters')

    try:
        js = json.loads(data)                                   #creates a dictionary or nested combination of dicts and lists named js
    except:
        js = None

    if not js or 'status' not in js or js['status'] != 'OK':           #if js is not true (if json is None then do ->), or the word status is not in the dict js, or status key does not have the value ok in it, do ->
        print('==== Failure To Retrieve ====')
        print(data)
        continue

    print(json.dumps(js, indent=2))                  #dumps is opp of loads, it'll print the dictionary with indent of 4

    lat = js['results'][0]['geometry']['location']['lat']
    lng = js['results'][0]['geometry']['location']['lng']
    print('lat', lat, 'lng', lng)
    location = js['results'][0]['formatted_address']
    print(location)